src/StoreApi/Utilities/CartController.php

```diff
@@ -213,17 +215,144 @@ public function validate_add_to_cart( \WC_Product $product, $request ) {
 		do_action( 'wooocommerce_store_api_validate_add_to_cart', $product, $request );
 	}
 
+	// phpcs:disable Squiz.Commenting.FunctionCommentThrowTag.WrongNumber -- this method _does_ throw a RouteException
+	// in convert_notices_to_exceptions.
 	/**
 	 * Validate all items in the cart and check for errors.
 	 *
-	 * @throws RouteException Exception if invalid data is detected.
+	 * @throws StockAvailabilityException Exception if invalid data is detected due to insufficient stock levels.
+	 * @throws RouteException Exception if invalid data is detected but not due to insufficient stock levels.
 	 */
 	public function validate_cart_items() {
 		$cart       = $this->get_cart_instance();
 		$cart_items = $this->get_cart_items();
 
+		$out_of_stock_products         = [];
+		$too_many_in_cart_products     = [];
+		$partial_out_of_stock_products = [];
+		$not_purchasable_products      = [];
+
 		foreach ( $cart_items as $cart_item_key => $cart_item ) {
-			$this->validate_cart_item( $cart_item );
+			try {
+				$this->validate_cart_item( $cart_item );
+			} catch ( TooManyInCartException $error ) {
+				$too_many_in_cart_products[] = $error;
+			} catch ( NotPurchasableException $error ) {
+				$not_purchasable_products[] = $error;
+			} catch ( PartialOutOfStockException $error ) {
+				$partial_out_of_stock_products[] = $error;
+			} catch ( OutOfStockException $error ) {
+				$out_of_stock_products[] = $error;
+			}
+		}
+
+		if (
+			count(
+				array_merge(
+					$out_of_stock_products,
+					$partial_out_of_stock_products,
+					$not_purchasable_products,
+					$too_many_in_cart_products
+				)
+			) > 0 ) {
+
+			$error = new WP_Error();
+
+			if ( count( $out_of_stock_products ) > 0 ) {
+				$out_of_stock_product_names = array_map(
+					function( $product ) {
+						return $product->getProductName();
+					},
+					$out_of_stock_products
+				);
+
+				$error->add(
+					409,
+					sprintf(
+						/* translators: %s: product names  */
+						_n(
+							'"%s" is out of stock and cannot be purchased.',
```

Comment:
I like what you've done with these messages but one change needed would be to wrap each product name with `""` rather than the full list. e.g. "Product 1 and Product 2" would become "Product 1" and "Product 2".

Reply:
Cheers, I've added this functionality to `natural_language_join`: https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/8094d9f98e7459e79d89c83eb116db9d82508fba/src/Utils/ArrayUtils.php#L17-L24